src/groupeddataframe/complextransforms.jl

```diff
@@ -92,9 +92,87 @@ function fill_row!(row, outcols::NTuple{N, AbstractVector},
+                                        i::Integer,
+                                        f::Base.Callable,
+                                        gd::GroupedDataFrame,
+                                        starts::AbstractVector{<:Integer},
+                                        ends::AbstractVector{<:Integer},
+                                        incols::Union{Nothing, AbstractVector,
+                                                      Tuple, NamedTuple},
+                                        colnames::NTuple{<:Any, Symbol},
+                                        firstmulticol::Val)
+    j = nothing
+    gdidx = gd.idx
+    local newoutcols
+    for i in idx
+        row = wrap_row(do_call(f, gdidx, starts, ends, gd, incols, i), firstmulticol)
+        j = fill_row!(row, outcols, i, 1, colnames)
+        if j !== nothing # Need to widen column
+            # If another thread is already widening outcols, wait until it's done
+            lock(widen_type_lock)
+            try
+                newoutcols = outcolsref[]
+                # If another thread widened outcols, try again
+                # to see whether new eltype is wide enough
+                if newoutcols !== outcols
+                    j = fill_row!(row, newoutcols, i, j, colnames)
+                end
+                while j !== nothing
+                    # Workaround for julia#15276
+                    newoutcols = let i=i, j=j, outcols=outcols, row=row, idx=idx
+                        ntuple(length(outcols)) do k
+                            S = typeof(row[k])
+                            T = eltype(outcols[k])
+                            U = promote_type(S, T)
+                            if S <: T || U <: T
+                                outcols[k]
+                            else
+                                copyto!(Tables.allocatecolumn(U, length(outcols[k])),
+                                        idx[1], outcols[k], idx[1], i - idx[1] + (k < j))
+                            end
+                        end
+                    end
+                    outcolsref[] = newoutcols
+                    j = fill_row!(row, newoutcols, i, j, colnames)
+                    type_widened .= true
+                    type_widened[tid] = false
+                end
+            finally
+                unlock(widen_type_lock)
+            end
+            return _combine_rows_with_first_task!(tid, idx, newoutcols, outcolsref,
+                                                  type_widened, widen_type_lock,
+                                                  i+1, f, gd, starts, ends,
+                                                  incols, colnames, firstmulticol)
+        end
+        # If other thread widened columns, copy already processed data to new vectors
+        # This doesn't have to happen immediately (hence type_widened isn't atomic),
+        # but the more we wait the more data will have to be copied
+        if type_widened[tid]
```

Comment:
the only question is if it is worth to put it here as type widening should be rare (but maybe it is OK as this test is cheap).

Reply:
Type widening itself is frequent. What is rare is the number of times we're going to widen columns. But for example with `Union{T, Missing}`, typically a `missing` result will be encountered relatively soon. If the number of groups is large, it makes sense to start writing results to the new columns right avoid to avoid copying everything at the end. That also allows doing that in parallel.

Reply:
Agreed.